### PR TITLE
Implemented server configurations

### PR DESCRIPTION
**Pull request by juanmuscaria**
Some observations and limitations about the implemented configuration options:
* Players do not receive knockback from explosions if it's damage is 0, setting the multiplier to a really low number works as a way to disable damage
* When "Allow Attachments" is set to false and the player tries to modify the attachments of a gun, all attachments will be removed
* Zombies attributes only applies to newly spawned zombies
* "Extra Reload Duration" does not change the duration of the animation itself, it might be something to be changed in the future

### CODE REVIEW

File: crafting-dead-core/src/main/java/com/craftingdead/core/world/action/RemoveMagazineAction.java [hunk 1]
```diff
@@ -97,7 +99,9 @@ protected void finish() {
       // This will be synced to the client by the gun.
       this.ammoProvider.setMagazineStack(ItemStack.EMPTY);
       if (!this.oldMagazineStack.isEmpty()
-          && this.getPerformer().getEntity() instanceof Player) {
+          && this.getPerformer().getEntity() instanceof Player
+          && !(this.oldMagazineStack.getCapability(Magazine.CAPABILITY).orElseThrow(() -> new IllegalStateException("No magazine capability")).isEmpty()
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/world/action/RemoveMagazineAction.java]: Why did you add a capability check here?

File: crafting-dead-core/src/main/java/com/craftingdead/core/world/action/reload/MagazineReloadAction.java [hunk 2]
```diff
@@ -72,7 +73,9 @@ protected void loadNewMagazineStack(boolean displayOnly) {
     this.ammoProvider.setMagazineStack(this.newMagazineStack);
     if (!displayOnly
         && !this.oldMagazineStack.isEmpty()
-        && this.getPerformer().getEntity() instanceof Player) {
+        && this.getPerformer().getEntity() instanceof Player
+        && !(this.oldMagazineStack.getCapability(Magazine.CAPABILITY).orElseThrow(() -> new IllegalStateException("No magazine capability")).isEmpty()
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/world/action/reload/MagazineReloadAction.java]: Why did you add a capability check here?

File: crafting-dead-core/src/main/java/com/craftingdead/core/world/entity/extension/LivingExtensionImpl.java [hunk 3]
```diff
@@ -359,8 +363,25 @@ private void updateClothing() {
...
+      if (CraftingDead.serverConfig.backstabEnable.get()) {
+        var usedMeleeWeapon = player.getItemInHand(player.getUsedItemHand())
+            .getItem() instanceof MeleeWeaponItem;
+        if (usedMeleeWeapon && !EntityUtil.canSee(this.getEntity(), player, 90F)) {
+          damage *= CraftingDead.serverConfig.backstabBonusDamage.get().floatValue();
+        }
+      }
+
+      if (CraftingDead.serverConfig.criticalHitsEnable.get()) {
+        if (CraftingDead.serverConfig.criticalHitsChance.get() > player.getRandom().nextDouble(100D)) {
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/world/entity/extension/LivingExtensionImpl.java]: You could use `Random#nextFloat`

File: crafting-dead-core/src/main/java/com/craftingdead/core/world/entity/grenade/FireGrenadeEntity.java [hunk 4]
```diff
@@ -60,19 +60,21 @@ public void activatedChanged(boolean activated) {
...
-            this.blockPosition().offset(FIRE_RADIUS, 0, FIRE_RADIUS)).forEach(blockPos -> {
-              if (this.level.getBlockState(blockPos).getBlock() == Blocks.AIR) {
-                if (Math.random() <= 0.8D) {
-                  this.level.setBlockAndUpdate(blockPos, Blocks.FIRE.defaultBlockState());
-                }
-              }
-            });
+        BlockPos.betweenClosedStream(this.blockPosition().offset(-fireRadius, 0, -fireRadius),
+            this.blockPosition().offset(fireRadius, 0, fireRadius)).forEach(blockPos -> {
+          if (this.level.getBlockState(blockPos).getBlock() == Blocks.AIR) {
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/world/entity/grenade/FireGrenadeEntity.java]: Use `BlockState#isAir` instead of `== Blocks.AIR`

File: crafting-dead-core/src/main/java/com/craftingdead/core/world/explosion/ConfigurableExplosion.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+import net.minecraft.world.level.ExplosionDamageCalculator;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.gameevent.GameEvent;
+import net.minecraft.world.level.material.FluidState;
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.Vec3;
+import org.jetbrains.annotations.Nullable;
+
+public class ConfigurableExplosion extends Explosion {
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/world/explosion/ConfigurableExplosion.java]: I'm not sure if you're familiar with Mixin but this would be much nicer as a Mixin instead of creating a new class and access transforming all the fields.

File: crafting-dead-survival/src/main/java/com/craftingdead/survival/CraftingDeadSurvival.java [hunk 6]
```diff
@@ -278,14 +282,95 @@ public void handleBiomeLoading(BiomeLoadingEvent event) {
...
+        }
+      }
+    }
+  }
+
+  @SubscribeEvent
+  public void handleLivingDrops(LivingDropsEvent event) {
+    if (event.getEntity() instanceof AdvancedZombie &&
+        !CraftingDeadSurvival.serverConfig.zombiesDeathDrops.get()) {
+      event.getDrops().clear();
```
Comment (josephtarbit) [inline crafting-dead-survival/src/main/java/com/craftingdead/survival/CraftingDeadSurvival.java]: Move this into `LivingExtensionImpl#handleDeathLoot`

File: crafting-dead-survival/src/main/java/com/craftingdead/survival/CraftingDeadSurvival.java [hunk 7]
```diff
@@ -278,14 +282,95 @@ public void handleBiomeLoading(BiomeLoadingEvent event) {
...
+    }
+  }
+
+  @SubscribeEvent
+  public void handleEntityJoinWorld(EntityJoinWorldEvent event) {
+    if (!event.getWorld().isClientSide() && !event.loadedFromDisk()) {
+      if (event.getEntity() instanceof AdvancedZombie entity) {
```
Comment (josephtarbit) [inline crafting-dead-survival/src/main/java/com/craftingdead/survival/CraftingDeadSurvival.java]: Override `finalizeSpawn` in `AdvancedZombie` and configure the attributes there.

File: crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+import net.minecraft.world.level.Explosion.BlockInteraction;
+import net.minecraftforge.common.ForgeConfigSpec;
+
+public class ServerConfig {
+
+  // ================================================================================
+  // Burst-fire Values
+  // ================================================================================
+
+  public final ForgeConfigSpec.BooleanValue burstfireEnable;
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java]: Use `enabled` instead of `enable` for variable names, e.g. rename this to `burstfireEnabled`

File: crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+  // ================================================================================
+  // Reload Values
+  // ================================================================================
+
+// TODO: Implement those once individual ammo/bullets are added - juanmuscaria
+//  public final ForgeConfigSpec.BooleanValue reloadReloadBulletsIndividually;
+//  public final ForgeConfigSpec.BooleanValue reloadTakeAmmoOnReload;
+//  public final ForgeConfigSpec.BooleanValue reloadTakeAmmoAsMagazine;
+  public final ForgeConfigSpec.BooleanValue reloadGunComeEmptyMag;
+  public final ForgeConfigSpec.IntValue reloadReloadDuration;
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java]: Remove duplicate `reload` in variable name, so rename to just `reloadDuration`

File: crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+// TODO : Implement those once dual wielding is added - juanmuscaria
+//  public final ForgeConfigSpec.IntValue reloadDualWieldSingleReloadDuration;
+//  public final ForgeConfigSpec.BooleanValue reloadDualWieldSoundsSingleReload;
+//  public final ForgeConfigSpec.BooleanValue reloadDualWieldSoundsShootWithNoAmmo;
+
+  // ================================================================================
+  // Scope Values
+  // ================================================================================
+
+  public final ForgeConfigSpec.BooleanValue scopeAllowAttachments;
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java]: Rename to `scopeAttachmentsAllowed`

File: crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+// TODO: Implement when pre define attachments are added - juanmuscaria
+//  public final ForgeConfigSpec.BooleanValue scopeAttachmentOverride;
+// TODO: Implement those when Night/Thermal vision scopes are added - juanmuscaria
+//  public final ForgeConfigSpec.BooleanValue scopeNightVision;
+//  public final ForgeConfigSpec.BooleanValue scopeThermalVision;
+  public final ForgeConfigSpec.DoubleValue scopeZoomAmountMultiplier;
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java]: Rename to `scopeZoomMultiplier`

File: crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+  public final ForgeConfigSpec.BooleanValue backstabEnable;
+  public final ForgeConfigSpec.DoubleValue backstabBonusDamage;
+
+  // ================================================================================
+  // Critical Hits Values
+  // ================================================================================
+
+  public final ForgeConfigSpec.BooleanValue criticalHitsEnable;
+  public final ForgeConfigSpec.DoubleValue criticalHitsBonusDamage;
+  public final ForgeConfigSpec.DoubleValue criticalHitsChance;
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java]: Change `hits` to `hit`, e.g. rename to `criticalHitChance`

File: crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java [hunk 13]
```diff
@@ -0,0 +1,468 @@
...
+        .push("Burstfire");
+    {
+      this.burstfireEnable = builder
+          .translation("options.craftingdead.server.burstfire.enable")
+          .comment("Enable (some) guns to fire a burst of shots")
+          .define("Enable", true);
+      this.burstfireShotsPerBurst = builder
+          .translation("options.craftingdead.server.burstfire.shots_per_burst")
+          .comment("The amount of shots per fired burst")
+          .defineInRange("Shots Per Burst", 3, 1, Integer.MAX_VALUE);}
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java]: Rename all config names to be the same as the field names, so change `Shots Per Burst` to `burstfireShotsPerBurst`

**Review by josephtarbit — CHANGES_REQUESTED** [review]

File: crafting-dead-core/src/main/java/com/craftingdead/core/world/action/RemoveMagazineAction.java (hunk 1 above)
Comment (juanmuscaria) [reply crafting-dead-core/src/main/java/com/craftingdead/core/world/action/RemoveMagazineAction.java]: It's to check if the magazine is empty for the destroy empty magazine configuration, though I should have used map and orElse(true), I'll be changing that.

File: crafting-dead-core/src/main/java/com/craftingdead/core/world/action/reload/MagazineReloadAction.java (hunk 2 above)
Comment (juanmuscaria) [reply crafting-dead-core/src/main/java/com/craftingdead/core/world/action/reload/MagazineReloadAction.java]: It's to check if the magazine is empty for the destroy empty magazine configuration, though I should have used map and orElse(true), I'll be changing that.

File: crafting-dead-core/src/main/java/com/craftingdead/core/world/entity/extension/LivingExtensionImpl.java (hunk 3 above)
Comment (juanmuscaria) [reply crafting-dead-core/src/main/java/com/craftingdead/core/world/entity/extension/LivingExtensionImpl.java]: Forge configs are double by default, so I used double to avoid any unnecessary cast

File: crafting-dead-core/src/main/java/com/craftingdead/core/world/explosion/ConfigurableExplosion.java (hunk 5 above)
Comment (juanmuscaria) [reply crafting-dead-core/src/main/java/com/craftingdead/core/world/explosion/ConfigurableExplosion.java]: I avoided changing vanilla behaviour just to be able to extend the grenade functionalities. But if necessary I can create a mixin

File: crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java (hunk 13 above)
Comment (juanmuscaria) [reply crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java]: My plan was to make something easy to read for anyone changing configurations, so I followed this rule for naming configurations:
* Each field will be named by <category+subcategories><short configuration name>
* Each configuration key have a self explanatory name
Resulting on the following config
```toml
#Comment
[Burstfire]
	#Comment
	Enable = true
	#Comment
	#Range: > 1
	"Shots Per Burst" = 3
```
compared to 
```toml
#Comment
[Burstfire]
	#Comment
	burstfireEnable = true
	#Comment
	#Range: > 1
	burstfireShotsPerBurst = 3
```

File: crafting-dead-core/src/main/java/com/craftingdead/core/world/explosion/ConfigurableExplosion.java (hunk 5 above)
Comment (josephtarbit) [reply crafting-dead-core/src/main/java/com/craftingdead/core/world/explosion/ConfigurableExplosion.java]: Add a check in the Mixin to see if the source entity is a grenade

File: crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java (hunk 13 above)
Comment (josephtarbit) [reply crafting-dead-core/src/main/java/com/craftingdead/core/ServerConfig.java]: I agree it's easier to read but it's a less universal format, as a compromise you could remove the category prefix but keep it camel case, such as `shotsPerBurst = 3`

File: crafting-dead-survival/src/main/java/com/craftingdead/survival/CraftingDeadSurvival.java (hunk 6 above)
Comment (juanmuscaria) [reply crafting-dead-survival/src/main/java/com/craftingdead/survival/CraftingDeadSurvival.java]: It's not possible due to "LivingExtensionImpl" being part of core and zombies part of survival.

Comment (josephtarbit) [reply crafting-dead-survival/src/main/java/com/craftingdead/survival/CraftingDeadSurvival.java]: See `com.craftingdead.survival.world.entity.SurvivalPlayerHandler`. You could create a class called `SurvivalLivingHandler` that implements `LivingHandler` which would allow you `handleDeathLoot`.
